crates/bin/prove_block/src/main.rs

```diff
@@ -321,146 +387,113 @@ async fn main() -> Result<(), Box<dyn Error>> {
-                log::debug!("contract class for {}: {:?}", contract_address, contract_class);
-
-                match contract_class {
-                    starknet::core::types::ContractClass::Sierra(flattened_sierra_cc) => {
-                        let middle_sierra: types::MiddleSierraContractClass = {
-                            let v = serde_json::to_value(flattened_sierra_cc).unwrap();
-                            serde_json::from_value(v).unwrap()
-                        };
-                        let sierra_cc = cairo_lang_starknet_classes::contract_class::ContractClass {
-                            sierra_program: middle_sierra.sierra_program,
-                            contract_class_version: middle_sierra.contract_class_version,
-                            entry_points_by_type: middle_sierra.entry_points_by_type,
-                            sierra_program_debug_info: None,
-                            abi: None,
-                        };
-                        let casm_cc =
-                            cairo_lang_starknet_classes::casm_contract_class::CasmContractClass::from_contract_class(
-                                sierra_cc.clone(),
-                                false,
-                                usize::MAX,
-                            )
-                            .unwrap();
-                        let casm_cc: GenericCasmContractClass = casm_cc.into();
-
-                        log::debug!("class_hash (from RPC): {}", class_hash);
-
-                        // TODO: it seems that this ends up computing the wrong class hash...
-                        write_class_facts(
-                            sierra_cc,
-                            casm_cc.clone(),
-                            &mut initial_state.ffc.clone_with_different_hash::<PoseidonHash>(),
-                        )
-                        .await?;
-
-                        compiled_classes.insert(class_hash, casm_cc);
-                    }
-                    starknet::core::types::ContractClass::Legacy(_compressed_logacy_contract_class) => {
-                        panic!("legacy class (TODO)");
-                    }
-                }
-            } else {
-                log::warn!("No class available for contract {}", contract_address);
+            let generic_cc = generic_sierra_cc.compile()?;
+
+            log::debug!("============================");
+            log::debug!("class_hash (from RPC): {:x?}", class_hash);
+            log::debug!("class_hash (computed): {:x?}", generic_sierra_cc.class_hash()?);
+
+            // TODO: it seems that this ends up computing the wrong class hash...
+            let (contract_class_hash, compiled_contract_hash) = write_class_facts(
+                generic_sierra_cc,
+                generic_cc.clone(),
+                &mut initial_state.ffc.clone_with_different_hash::<PoseidonHash>(),
```

Comment:
Why do we need to use a different hash function?

Reply:
Because the class trie uses Poseidon hashing whereas the storage trie uses Pedersen hashing.

But -- `SharedState` has a `ffc_for_class_hash` which would be better to use here (TODO)

Reply:
Changed to `ffc_for_class_hash` in https://github.com/Moonsong-Labs/snos/pull/19/commits/8575e0822b76949eac25112046d7f061727c364d